feat(articleHitsCount): 게시글 조회수 기능 추가

JoonBro opened the pull request:
조회수 API 없이 
게시글 조회할 때 -> increase 함수 수행 -> 첫 조회일때 객체 생성 후 increase
게시글 삭제할때 -> articleHitsCount 객체 삭제

위 로직으로 구현하였습니다.

게시글 조회, 삭제할 때 컨트롤러단에서 서비스를 참조하여 수행해야하는지, 서비스 함수 내에서 해야하는지 잘 몰라서 일단 articleService 내에 구현하였습니다.

조회수라 decrease 만들었다가 필요없을거 같아서 삭제했습니다.

Commit:
articleHitsCount 구현

Commit:
articleHitsCount decrease function delete

hungryjayy commented on src/main/kotlin/com/devfloor/untitled/articlehitscount/domain/ArticleHitsCountRepository.kt:
```diff
@@ -0,0 +1,17 @@
+
+@Transactional(readOnly = true)
+interface ArticleHitsCountRepository : JpaRepository<ArticleHitsCount, Long> {
+    fun findByArticleOrNull(article: Article): ArticleHitsCount?
```
혹시 `findBy~~()` 이거는 `Optional<T>`  을 반환하는 것 아닌가요!?
따라서 HashtagRepository에 구현되어있는 `findByNameOrNull` 처럼 해야하지 않나!? 라는 생각이 들었어요.

그리고 Null이라는걸 JpaRepository가 해석하지 못하는 거 같던데 아닌가요 !? "Cannot resolve property 'null' " 라는 문구가 보이네요 !

JoonBro replied:
네 수정하겠습니당!

hungryjayy commented on src/main/kotlin/com/devfloor/untitled/articlehitscount/application/ArticleHitsCountService.kt:
```diff
@@ -0,0 +1,22 @@
+) {
+    @Transactional
+    fun increase(article: Article) {
+        var articleHitsCount = articleHitsCountRepository.findByArticleOrNull(article)
+        if (articleHitsCount == null) {
+            articleHitsCount = ArticleHitsCount(article)
+        }
```
여기 엘비스 연산자를 이용하는건 어떨까요 !?
```suggestion
       var articleHitsCount = articleHitsCountRepository.findByArticleOrNull(article)
            ?: ArticleHitsCount(article)
```

JoonBro replied:
수정했습니다!

hungryjayy commented on src/main/kotlin/com/devfloor/untitled/articlehitscount/application/ArticleHitsCountService.kt:
```diff
@@ -0,0 +1,22 @@
+    private val articleHitsCountRepository: ArticleHitsCountRepository,
+) {
+    @Transactional
+    fun increase(article: Article) {
+        var articleHitsCount = articleHitsCountRepository.findByArticleOrNull(article)
+        if (articleHitsCount == null) {
+            articleHitsCount = ArticleHitsCount(article)
+        }
+        articleHitsCount.increase()
+        articleHitsCountRepository.save(articleHitsCount)
+    }
```
아니면 전체를 이러한 형태로 바꾸는건 어떨까요!?
조회해서 있으면 바꿔주고 더티체킹 이용하도록, 없으면 직접 저장해주는 로직으로 이렇게 짜보았는데, 문제는 이렇게 할라면 ArticleHits 생성 시점에 좋아요 값을 1로 생성하도록 해야할거같아요.!

```suggestion
    fun increase(article: Article) = articleHitsCountRepository.findByArticleOrNull(article)
        ?.run { ::increase }
        ?: articleHitsCountRepository.save(ArticleHitsCount(article))
```

hungryjayy replied:
suggestion 코드 14번째 줄에서 자동으로 개행이 되어 보이네요.. 의도한 바는 아래와 같습니다 !
![image](https://user-images.githubusercontent.com/49271821/122709580-89e13880-d299-11eb-8a4b-d22b7b0f3076.png)

JoonBro replied:
increase 이후 (?. run에 해당하는) 변경된 객체를 save해야하는데 increase 자체가 setter함수로 리턴값없이 구현해놔서..
setter함수의경우 해당 변경한 객체를 리턴하도록 만들어도 되는건가요?? 대부분 자료가 void로 구현되어지길래 ..

hungryjayy reviewed (commented):
일단 간단하게 남기긴 했는데, 
어제 이야기한 부분에 대해서 반영하신다면 그때 추가적으로 확인해보겠습니다.!!

Commit:
1. 명명 articleClickCount로 변경
2. findByArticleOrNull 수정 개발
3. articleService 내에 조회수 로직 추가

hungryjayy commented on src/main/kotlin/com/devfloor/untitled/articleclickcount/domain/ArticleClickCount.kt:
```diff
@@ -0,0 +1,29 @@
+import javax.persistence.OneToOne
+
+@Entity
+class ArticleClickCount(
+    article: Article,
+) {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "article_hits_id")
+    val id: Long = 0
+
+    @OneToOne
+    @JoinColumn(name = "article_id")
+    val article: Article = article
+
+    var hitsCount: Long = 1
+        protected set
+
+    fun increase() = this.hitsCount++
+}
```
저희 저번에 ClickCount로 명명을 정했었나요!?
ViewCount였는지 ClickCount였는지 헷갈리는데 개인적으로는 둘다 좋아보이네요 !

객체 명을 ArticleClickCount로 간 이상 프로퍼티 hitsCount도 네이밍을 객체와 통일하면 어떨까요 !?

JoonBro replied:
아 저도 기억이 잘 안나서 변수 바꾸는건 까먹었네요!! 저도 둘다 괜찮아보이는데 준영이 의견 보고 통일시키겠습니다!

lxxjn0 replied:
흔히 조회수를 view로 많이 사용하는 것 같아서 ArticleViewCount가 좋을 것 같은데 어때??

hungryjayy commented on src/main/kotlin/com/devfloor/untitled/articleclickcount/domain/ArticleClickCountRepository.kt:
```diff
@@ -0,0 +1,24 @@
+interface ArticleClickCountRepository : JpaRepository<ArticleClickCount, Long> {
+    fun findByArticle(article: Article): Optional<ArticleClickCount>
+
+    @Transactional
+    @Modifying(clearAutomatically = true, flushAutomatically = true)
+    @Query(value = "DELETE FROM ArticleClickCount WHERE article = :article")
+    fun deleteByArticle(article: Article)
```
Article과 ArticleClickCount의 관계는 일대일 아닌가요!?
만약 그렇게 되면 벌크 연산을 수행하지 않아도 되는 것 아닌가요??

JoonBro replied:
그러네요! 수정하겠습니다!

hungryjayy commented on src/main/kotlin/com/devfloor/untitled/article/application/ArticleService.kt:
```diff
@@ -108,6 +114,14 @@ class ArticleService(
         }
         ?: EntityNotFoundException.notExistsId(Article::class, articleId)
+
+    @Transactional
+    fun increaseHitCount(article: Article) {
+        articleClickCountRepository.findByArticleOrNull(article)
+            ?.run { ::increase }
+            ?: articleClickCountRepository.save(ArticleClickCount(article))
+    }
```
1. 여기도 네이밍 Click으로 통일하는게 어떨까요 !?
2. 게시글 조회시에만 이 메서드가 사용 될 것 같은데, private으로 바꾸면 어떨까요!?
3. 같은 맥락으로 게시글 조회에만 사용되기 때문에 이 메소드는 항상 `showByArticleId()`에서 호출될 것 같은데, 이 때 Transactional 어노테이션 자체가 필요없을 것 같습니다.! 더 중요한건, 만약 상위 메서드인 `showByArticleId()`에 Transactional 어노테이션이 붙어있지 않다고 하더라도 이 메서드는 트랜젝셔널하게 수행되지 않아요. 왜냐하면 프록시 객체에서 내부 메서드를 호출할 때에는 해당 프록시의 메서드를 Transactioanl로 감쌌는지 알 수가 없기 때문인거로 알고 있습니다 ! 만약 이거와 관련해서 런타임에 의도한 바와 다르게 수행된다면 아마 원인을 찾기가 어려울 것 같아요!
```suggestion
    private fun increaseClickCount(article: Article) {
        articleClickCountRepository.findByArticleOrNull(article)
            ?.run { ::increase }
            ?: articleClickCountRepository.save(ArticleClickCount(article))
    }
```

JoonBro replied:
1. 통일하겠습니다 놓쳤습니다.
2. private이 좋아보입니다! 수정하겠습니다.
3. 덕분에 transactional 어노테이션과 프록시객체에 관해서 공부좀했네용! 감사합니당. 같은 클래스 내 함수 / private 함수는 transactional 이 적용되지 않네용 !

hungryjayy commented on src/main/kotlin/com/devfloor/untitled/articleclickcount/domain/ArticleClickCount.kt:
```diff
@@ -0,0 +1,29 @@
+    @OneToOne
+    @JoinColumn(name = "article_id")
+    val article: Article = article
+
+    var hitsCount: Long = 1
+        protected set
```
이 프로퍼티 혹시 @Colume 을 사용하지 않아도 되는건가요!?

JoonBro replied:
수정하겠습니다!

hungryjayy reviewed (commented):
리뷰 남겼습니다 !

Commit:
1. viewCount로 명명 수정
2. increaseViewCount private처리, transactional 제거
3. viewCount @Column 처리
4. 삭제 벌크 연산 제거

lxxjn0 commented on src/main/kotlin/com/devfloor/untitled/article/application/ArticleService.kt:
```diff
@@ -108,6 +114,13 @@ class ArticleService(
         }
         ?: EntityNotFoundException.notExistsId(Article::class, articleId)
+
+    private fun increaseViewCount(article: Article) {
+        articleViewCountRepository.findByArticleOrNull(article)
+            ?.run { ::increase }
+            ?: articleViewCountRepository.save(ArticleViewCount(article))
+    }
```
이건 주원이형이 슬랙에서 이야기 나눈 것 같아서 체크만 해두겠씀~~~

JoonBro replied:
여기 문제 없어????

lxxjn0 commented on src/main/kotlin/com/devfloor/untitled/articleviewcount/domain/ArticleViewCount.kt:
```diff
@@ -0,0 +1,30 @@
+import javax.persistence.JoinColumn
+import javax.persistence.OneToOne
+
+@Entity
+class ArticleViewCount(
+    article: Article,
+) {
```
```suggestion
@Entity
@Table(name = "article_view_counts")
class ArticleViewCount(
    article: Article,
) {
```

테이블 어노테이션이 저렇게 쓰는게 맞았나? 잘 기억은 안나는데 다른 엔티티 참고해서 저것도 해두면 좋을 듯!!

JoonBro replied:
확인해볼게~

lxxjn0 commented on src/main/kotlin/com/devfloor/untitled/articleviewcount/domain/ArticleViewCount.kt:
```diff
@@ -0,0 +1,30 @@
+    @JoinColumn(name = "article_id")
+    val article: Article = article
+
+    @Column(name = "viewCount")
+    var viewCount: Long = 1
+        protected set
```
```suggestion
    @Column(name = "count")
    var count: Long = 1
        protected set
```

간결하게 count로 가져가는 것도 괜찮을 것 같은데 어때?

JoonBro replied:
오키~

lxxjn0 reviewed (approved):
크게 수정할 부분은 안보여서 이 정도만 수정하면 될 것 같어!! 👍🏼

hungryjayy reviewed (approved):
준영이가 남긴 부분만 수정하면 될거같아서 저는 Approve 하겠습니다 !! 고생하셨습니다~

Commit:
1. count 변수명 변경
2. table 어노테이션 적용

hungryjayy commented on src/main/kotlin/com/devfloor/untitled/articleviewcount/domain/ArticleViewCountRepository.kt:
```diff
@@ -0,0 +1,19 @@
+@Transactional(readOnly = true)
+interface ArticleViewCountRepository : JpaRepository<ArticleViewCount, Long> {
+    fun findByArticle(article: Article): Optional<ArticleViewCount>
+
+    fun deleteByArticle(article: Article)
```
다른 부분은 다 괜찮은 것 같고 이 부분 bulk연산 제거 관련 리뷰 반영하시면서 Transactional까지 제거된 것 같아요 !!
여기만 반영하시면 될거같습니다 !
```suggestion
    @Transactional
    fun deleteByArticle(article: Article)
```

Commit:
1. delete transactional 어노테이션 적용
